core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt

```diff
@@ -0,0 +1,199 @@
+      setInitialScale(INITIAL_SCALE)
+      builtInZoomControls = true
+      displayZoomControls = false
+      clearCache(true)
+      allowUniversalAccessFromFileURLs = true
+      setWebViewClient(webViewClient)
+    }
+    webChromeClient = KiwixWebChromeClient(callback, nonVideoView, videoView, this).apply {
+      setOnToggledFullscreen { fullscreen ->
+        setWindowVisibility(
+          if (fullscreen) View.SYSTEM_UI_FLAG_LOW_PROFILE else View.SYSTEM_UI_FLAG_VISIBLE
+        )
+      }
+    }
+  }
+
+  override fun performLongClick(): Boolean {
+    val result = hitTestResult
+    if (result.type == HitTestResult.SRC_ANCHOR_TYPE) {
+      callback.webViewLongClick(result.extra)
+      return true
+    }
+    return super.performLongClick()
+  }
+
+  override fun onCreateContextMenu(menu: ContextMenu) {
+    super.onCreateContextMenu(menu)
+    val result = hitTestResult
+    if (result.type == HitTestResult.IMAGE_ANCHOR_TYPE ||
+      result.type == HitTestResult.IMAGE_TYPE ||
+      result.type == HitTestResult.SRC_IMAGE_ANCHOR_TYPE
+    ) {
+      val saveMenu =
+        menu.add(0, 1, 0, resources.getString(R.string.save_media))
+      saveMenu.setOnMenuItemClickListener {
+        val msg = SaveHandler(zimReaderContainer).obtainMessage()
+        requestFocusNodeHref(msg)
+        true
+      }
+    }
+  }
+
+  override fun onAttachedToWindow() {
+    super.onAttachedToWindow()
+    compositeDisposable.add(
+      sharedPreferenceUtil.textZooms
+        .subscribe {
+          settings.textZoom = it
+        }
+    )
+  }
+
+  override fun onDetachedFromWindow() {
+    super.onDetachedFromWindow()
+    compositeDisposable.clear()
+  }
+
+  override fun onScrollChanged(l: Int, t: Int, oldl: Int, oldt: Int) {
+    super.onScrollChanged(l, t, oldl, oldt)
+    val windowHeight = if (measuredHeight > 0) measuredHeight else 1
+    val pages = contentHeight / windowHeight
+    val page = t / windowHeight
+    callback.webViewPageChanged(page, pages)
+  }
+
+  internal class SaveHandler(private val zimReaderContainer: ZimReaderContainer) :
+    Handler() {
+    private fun getDecodedFileName(url: String?, src: String?): String =
+      url?.substringAfterLast("/", "")
+        ?.takeIf { it.contains(".") }
+        ?: src?.substringAfterLast("/", "")
+          ?.substringAfterLast("%3A") ?: ""
+
+    override fun handleMessage(msg: Message) {
+      val url = msg.data["url"] as? String
+      val src = msg.data["src"] as? String
+      if (url != null || src != null) {
+        val fileName = getDecodedFileName(url, src)
+        val dotIndex = fileName.lastIndexOf('.')
+        var root =
+          Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES)
+        if (instance.externalMediaDirs.isNotEmpty()) {
+          root = instance.externalMediaDirs[0]
+        }
+        var storageDir = File(root, fileName)
+        var newUrl = fileName
+        var i = 2
+        while (storageDir.exists()) {
```

Comment:
we could make this a whole lot less confusing. I think this is a fun use case for a sequence

Reply:
Should I make comment for this?

Reply:
I would like for you to read about [sequences](https://kotlinlang.org/docs/reference/sequences.html) (because they are cool and useful)) and think about how this loop could be rewritten so it is actually somewhat understandable. If you want more guidance ask away

Reply:
```
  var storageDir = File(root, fileName)
        var newUrl = fileName
        var i = 2
        generateSequence {
          newUrl = fileName.substring(0, dotIndex) + "_" + i + fileName.substring(dotIndex)
          storageDir = File(root, newUrl)
          i++
        }.takeWhile { storageDir.exists() }
```
@macgills Is this the more readable way?

Reply:
what you want to do is find a file that does not exist and use that to write the image contents to.

For an example file "File("directory", "image.png")" we should use that if that file does not exist
If that file exists then we want to enumerate file names until we get one that does not exist and then use that, a potential output would be `File("directory", "image_1.png")`

I would start the chain of operations with a `generateSequence(1) { it+1 }`, the end of this sequence should return a File that does not exist yet

Reply:
```
 if (storageDir.exists()) {
           generateSequence(2) { it + 1 }.takeWhile {
             newUrl = fileName.substring(0, dotIndex) + "_" + it + fileName.substring(dotIndex)
             storageDir = File(root, newUrl)
             storageDir.exists()
           }
         }
```
and 
```
   if (storageDir.exists()) {
          generateSequence(1) {
            newUrl = fileName.substring(0, dotIndex) + "_" + it + fileName.substring(dotIndex)
            storageDir = File(root, newUrl)
            if (storageDir.exists()) return@generateSequence it + 1 else null
          }
        }
```
@macgills I tried the above approaches but it didn't work. I don't know what's wrong with this code?

Reply:
You want to create a sequence of files and then find one that doesn't exist, let me highlight some methods on sequences which are very similar to methods on list

![image](https://user-images.githubusercontent.com/3358282/93066310-bebcc100-f671-11ea-92a0-b144c4dcbbd2.png)

![image](https://user-images.githubusercontent.com/3358282/93066698-37bc1880-f672-11ea-940d-03ed689b22e9.png)

Explore the API, paying special attention to the return types
![image](https://user-images.githubusercontent.com/3358282/93067330-0132cd80-f673-11ea-9320-f967eecaac85.png)

This should set you on the right path, feel free to message me on slack if you need further assistance